### PR TITLE
feat(common/models): naive correction-algorithm timer

### PR DESCRIPTION
**Pull request by jahorton**
To keep things snappy, this adds a (by default) 33ms timer on the correction algorithm.  It will automatically return results after that time, whether or not any valid corrections have been found.

Note: this time will not be cut short if/when a new keystroke is received.  That's why I'm labelling it 'naive'.  See #3580 for notes toward a more robust (but also markedly more complex) design... the 'complex' part being why we're going 'naive' for now.

### CODE REVIEW

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 1]
```diff
@@ -262,13 +259,31 @@ namespace correction {
...
+  type PathResult = NullPath | IntermediateSearchPath | CompleteSearchPath;
+
   // The set of search spaces corresponding to the same 'context' for search.
   // Whenever a wordbreak boundary is crossed, a new instance should be made.
   export class SearchSpace {
     private QUEUE_SPACE_COMPARATOR: models.Comparator<SearchSpaceTier>;
 
     static readonly EDIT_DISTANCE_COST_SCALE = 5;
     static readonly MIN_KEYSTROKE_PROBABILITY = 0.0001;
+    static readonly DEFAULT_ALLOTTED_CORRECTION_TIME_INTERVAL = 33; // in milliseconds.
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:286]: 👍

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 2]
```diff
@@ -406,94 +418,116 @@ namespace correction {
...
       // might should also include a 'base cost' parameter of sorts?
       let searchSpace = this;
       let currentReturns: {[mapKey: string]: SearchNode} = {};
 
+      // JS measures time by the number of milliseconds since Jan 1, 1970.
+      let timeStart = Date.now();
+      let maxTime: number;
+      if(waitMillis == 0) {
+        maxTime = Infinity;
+      } else if(waitMillis == undefined || Number.isNaN(waitMillis)) { // also covers null.
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:525]: I'm a bit confused here about the "also covers null" comment. What is `Number.isNaN(null)` trying to accomplish?

suppose `waitMills = null`:

Try this in a JavaScript console:

```javascript
> null == undefined
true
> Number.isNaN(null)
true
> waitMillis == undefined || Number.isNaN(waitMillis)
true
> null == undefined || Number.isNaN(null)
true
> true || false
true
```

File: common/predictive-text/worker/correction/distance-modeler.ts [hunk 3]
```diff
@@ -262,13 +259,31 @@ namespace correction {
...
+    cost: number
+  }
+
+  type CompleteSearchPath = {
+    type: 'complete',
+    cost: number,
+    finalNode: SearchNode
+  }
+
+  type PathResult = NullPath | IntermediateSearchPath | CompleteSearchPath;
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/distance-modeler.ts:277]: Nice discriminated union!

**Review by eddieantonio — APPROVED** [review]
I have one thing I'm confused about, but other than that, LGTM!

File: common/predictive-text/worker/correction/distance-modeler.ts (hunk 2 above)
Comment (jahorton) [reply common/predictive-text/worker/correction/distance-modeler.ts:525]:
```
> Number.isNaN(undefined)
false
> undefined == NaN
false
> null == NaN
false
```

Comment (eddieantonio) [reply common/predictive-text/worker/correction/distance-modeler.ts:525]: Oh, so the `waitMillis == undefined` is what the comment is referring to! I had thought the comment was referring to `Number.isNaN(waitMillis)`. Haha, yeah, I already new about the `null == undefined` trick! https://gist.github.com/eddieantonio/2664cc5902516035949b732bbe19ac2b

**Pull request merged**